File: product-of-array-except-self.py
class Solution:
    # O(n), O(n)
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        answer = []
        product = 1
        
        for i in range(len(nums)):
            answer.append(product)
            product *= nums[i]
        
        product = 1
        
        for i in range(len(nums) - 1, 0 - 1, -1):
            answer[i] *= product
            product *= nums[i]
        
        return answer
    
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        product = 1
        zeros = nums.count(0)
        if zeros > 2:
            return [0 for _ in nums]
        elif zeros == 1:
            index = nums.index(0)
            for num in nums:
                if num != 0:
                    product *= num
            return [product if index == idx else 0 for idx, _ in enumerate(nums)]
        else:
            answer = []
            for num in nums:
                answer.append(product)
                product *= num
            product = 1
            for idx, num in enumerate(nums[::-1]):
                answer[-(idx+1)] *= product
                product *= num
            return answer

    # with division
    def productExceptSelf2(self, nums: List[int]) -> List[int]:
        product_all = 1
        zeros = nums.count(0)
        if zeros == 0:
            for num in nums:
                product_all *= num
            return [product_all // num for num in nums]
        elif zeros == 1:
            for num in nums:
                if num != 0:
                    product_all *= num
            index = nums.index(0)
            return [product_all if idx == index else 0 for idx, _ in enumerate(nums)]
        else:
            return [0 for _ in nums]

File: product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] left = new int[nums.length + 1];
        int[] right = new int[nums.length + 1];
        int[] answer = new int[nums.length];

        left[0] = 1;
        for(int i=0; i<nums.length; i++) {
            left[i + 1] = nums[i] * left[i];
        }
        right[nums.length] = 1;
        for(int i=nums.length; i>0; i--) {
            right[i - 1] = nums[i - 1] * right[i];
        }
        for(int i=0; i<nums.length; i++) {
            answer[i] = left[i] * right[i + 1];
        }

        return answer;
    }
}